cmd/collector/app/zipkin/http_handler_test.go

```diff
@@ -225,7 +226,7 @@ func TestFormatBadBody(t *testing.T) {
 	statusCode, resBodyStr, err := postBytes(server.URL+`/api/v1/spans`, []byte("not good"), createHeader("application/x-thrift"))
 	assert.NoError(t, err)
 	assert.EqualValues(t, http.StatusBadRequest, statusCode)
-	assert.EqualValues(t, "Unable to process request body: Unknown data type 111\n", resBodyStr)
```

Comment:
This change was interesting, as the new error is this: `Unable to process request body: size exceeded max allowed: 1869881447`. This seems to indicate that the fix is being picked up, but it did raise a warning in my head: are we really allocating more than 1GiB for this? Apparently, no. I used `runtime.MemStats` to measure the memory consumption before and after the HTTP call, and the usage was minimal (2 Alloc vs. 3 Alloc).

Reply:
This is indeed interesting.

The old error message "Unknown data type" comes from [`Skip` function](https://github.com/apache/thrift/blob/4abc5cfb4b3dea54484ca49a584b472a7373f3c5/lib/go/thrift/protocol.go#L177), which is called when either the field id is not defined in the struct, or when the type of the field id doesn't match what's defined in the struct.

The new error message "size exceeded max allowed" comes from [size sanity checks](https://github.com/apache/thrift/blob/4abc5cfb4b3dea54484ca49a584b472a7373f3c5/lib/go/thrift/configuration.go#L315), which could be called by a lot of functions (for example, `ReadString`, `ReadBinary`, `ReadMapBegin`, `ReadListBegin`, `ReadSetBegin`). This means the new code actually passed the field type check and is no longer skipped (or maybe previously the first field happened to match and it's the second field failed the field type check and caused the error, while in the new code the first field passed field type check but failed the size sanity check).

`runtime.MemStats` might be misleading. I believe it's the same as benchmark tests' `ReportAllocs`, and this benchmark test reported 0 allocs/0 bytes which is certainly a lie (I have to use a much smaller size because the old one is too slow):
```go
package main

import (
        "testing"
)

// const size = 1869881447
const size = 1869

func BenchmarkAlloc(b *testing.B) {
        b.ReportAllocs()
        for i := 0; i < b.N; i++ {
                _ = make([]int, size)
        }
}
```
```
$ go test -bench . -benchmem
goos: linux
goarch: amd64
pkg: foo
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkAlloc-12       1000000000               0.2487 ns/op          0 B/op          0 allocs/op
PASS
ok      foo     0.278s
```

Reply:
@fishy, do you think this change raises flags? I wouldn't expect the payload for this test to cause such a huge message.

Reply:
>do you think this change raises flags?

Not necessarily.

If you dig into the code on what this does, the first ever read from the thrift payload is to [`ReadListBegin`](https://github.com/jaegertracing/jaeger/blob/ff32436682897b1cf83c256716c0f45f67b3b595/model/converter/thrift/zipkin/deserialize.go#L46), which is one of the functions that would trigger the new error.

Let's first convert the payload (`[]byte("not good")`) into raw bytes: `[6e 6f 74 20 67 6f 6f 64]`.

In [`ReadListBegin` implementation](https://github.com/apache/thrift/blob/29fb3464c948f26474a69cf345d29584f15513a0/lib/go/thrift/binary_protocol.go#L366), first it reads a byte, `6e`, then it tries to read an int32 as the size of the list. For the reading of the int32, it reads the next 4 bytes (`6f 74 20 67`), decode with bigendian, which [results in 1869881447](https://play.golang.org/p/X35mEqoBHNA)

So this is totally expected behavior.

Reply:
And following `ReadListBegin`, jaeger's code didn't really do the pre-allocation. it just append spans to the slice and rely on `append` to do the allocation and grow the slice.

Reply:
y'all even already have a comment on that :)

https://github.com/jaegertracing/jaeger/blob/ff32436682897b1cf83c256716c0f45f67b3b595/model/converter/thrift/zipkin/deserialize.go#L51-L52